src/mame/sinclair/spec_snqk.cpp

```diff
@@ -2397,6 +2758,124 @@ void spectrum_state::setup_z80(uint8_t *snapdata, uint32_t snapsize)
 	}
 }
 
+/*
+	Load a .SPG (Spectrum Prog) file.
+
+	v1.1: https://raw.githubusercontent.com/tslabs/zx-evo/master/pentevo/docs/Formats/SPGv1_1.txt
+	v1.0: https://raw.githubusercontent.com/tslabs/zx-evo/master/pentevo/docs/Formats/SPGv1_0.txt
+	v0.2  https://raw.githubusercontent.com/tslabs/zx-evo/master/pentevo/docs/Formats/SPGv0_2.txt
+*/
+void spectrum_state::setup_spg(u8 *snapdata, u32 snapsize)
+{
+	struct spg_v10
+	{
+		char author[32];
+		char magic[12];
+		struct
+		{
+			u8 minor : 4;
+			u8 major : 4;
+		} version;
+		struct
+		{
+			u8 day;
+			u8 month;
+			u8 year; // 2000 + X
+		} date;
+		u16 pc;
+		u16 sp;
+		u8 win3_pg;
+		struct
+		{
+			u8 clock : 2;
+			u8 ei : 1;
+			u8 reserved : 5;
+		} cpu;
+		u16 pager_address;
+		u16 resident_address;
+		u16 blocks_num;
+		struct
+		{
+			u8 seconds;
+			u8 minutes;
+			u8 hours;
+		} time;
+		u8 reserved_1[17];
+		char utility[32];
+		u8 reserved_2[144];
+		struct block
+		{
+			struct
+			{
+				u8 address512 : 5; // 0-#C000, 1f-#FE00
+				u8 reserved_1 : 2;
+				bool is_last : 1;
+			};
+			struct
+			{
+				u8 size512 : 5; // 0-512b, 1f-16kB
+				u8 reserved_2 : 1;
+				u8 compression : 2; // 0-NONE, 1-MLZ, 2-HRUST
+			};
+			u8 page; // #00-#DF
+		} blocks[256];
+		u8 data;
+	};
+
+	spg_v10* spg = (spg_v10*)snapdata;
```

Comment:
It's not possible because declared blocks array will always oversize real data. If that can cause problems I can separate blocks from struct and process them iteratively

Reply:
@cuavas sorry to bother... what is the preference here?